sno/conflicts.py

```diff
@@ -12,6 +16,157 @@
 L = logging.getLogger("sno.conflicts")
 
 
+class ConflictIndex:
+    """
+    Like a pygit2.Index, but every conflict has a short key independent of its path,
+    and the entire index including conflicts can be serialised to a tree.
+    Conflicts are easier to modify than in a pygit2.Index (where they are backed by C iterators).
+    When serialised to a tree, conflicts will be added in a special .conflicts/ directory.
+    """
+
+    # We could use pygit2.IndexEntry everywhere but it has unhelpful __eq__ and __repr__ behaviour.
+    # So we have this equivalent struct.
+    Entry = namedtuple("Entry", ("path", "id", "mode"))
+
+    def __init__(self, index, serialised=False):
+        self.entries = {}
+        self.conflicts = {}
+
+        if serialised and index.conflicts:
+            raise RuntimeError(
+                "pygit2.Index.conflicts should be empty if index has been serialised"
+            )
+
+        for entry in index:
+            if entry.path.startswith(".conflicts/"):
+                if not serialised:
+                    raise RuntimeError(
+                        ".conflicts/ directory shouldn't exist if index has not been serialised"
+                    )
+                key, conflict_part = self._deserialise_conflict_part(entry)
+                self.conflicts.setdefault(key, AncestorOursTheirs.EMPTY)
+                self.add_conflict(key, self.conflicts.get(key) | conflict_part)
+            else:
+                self.add(entry)
+
+        if index.conflicts:
+            for key, conflict3 in enumerate(index.conflicts):
+                self.add_conflict(str(key), conflict3)
+
+    def __eq__(self, other):
+        if not isinstance(other, ConflictIndex):
+            return False
+        return self.entries == other.entries and self.conflicts == other.conflicts
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
```

Comment:
This method isn't required since python 3.x, `__ne__` delegates to `__eq__` by default

Reply:
Done.